core/core.go

```diff
@@ -1,189 +1,69 @@
-func NewPlantUMLClient(optFns ...func(*options)) PlantUMLClient {
-	o := options{
-		httpClient: nil,
-	}
-
-	for _, fn := range optFns {
-		fn(&o)
-	}
-
-	resolveHTTPClient(&o)
-
-	return &client{
-		options: o,
-		baseURL: baseURL,
-	}
+// ResponseDiagram response object.
+type ResponseDiagram interface {
+	// ToJSON encodes the result as JSON.
+	ToJSON() []byte
 }
 
-func resolveHTTPClient(o *options) {
-	if o.httpClient == nil {
-		o.httpClient = &http.Client{Timeout: defaultTimeout}
-	}
+// ResponseC4Diagram resulting C4 diagram.
+type ResponseC4Diagram struct {
+	SVG string `json:"svg"`
 }
 
-// code2Path converts the diagram as code to the 64Bytes encoded string to query plantuml
-//
-// Example: the diagram's code
-// @startuml
-//
-//	a -> b
-//
-// @enduml
-//
-// will be converted to SoWkIImgAStDuL80WaG5NJk592w7rBmKe100
-//
-// The resulting string to be used to generate C4 diagram
-// - as png: GET www.plantuml.com/plantuml/png/SoWkIImgAStDuL80WaG5NJk592w7rBmKe100
-// - as svg: GET www.plantuml.com/plantuml/svg/SoWkIImgAStDuL80WaG5NJk592w7rBmKe100
-func code2Path(s string) (string, error) {
-	zb, err := compress(*(*[]byte)(unsafe.Pointer(&s)))
+func (r ResponseC4Diagram) ToJSON() []byte {
+	// FIXME(?): add svg validation.
+	o, err := json.Marshal(r)
 	if err != nil {
-		return "", err
+		panic(err)
 	}
-	return encode64(zb), nil
+	return o
 }
 
-func compress(v []byte) ([]byte, error) {
-	var options = compression.DefaultOptions()
-	var w bytes.Buffer
-	if err := compression.Compress(&options, compression.FORMAT_DEFLATE, v, &w); err != nil {
-		return nil, err
-	}
-	return w.Bytes(), nil
+// ClientInputToGraph client to convert user input inquiry to the DiagramGraph.
+type ClientInputToGraph interface {
+	Do(string) (DiagramGraph, error)
 }
 
-func encode64(e []byte) string {
-	var r bytes.Buffer
-	for i := 0; i < len(e); i += 3 {
-		switch len(e) {
-		case i + 2:
-			r.Write(append3bytes(e[i], e[i+1], 0))
-		case i + 1:
-			r.Write(append3bytes(e[i], 0, 0))
-		default:
-			r.Write(append3bytes(e[i], e[i+1], e[i+2]))
-		}
-	}
-	return r.String()
+// ClientGraphToDiagram client to convert DiagramGraph to diagram artifact, e.g. svg image.
+type ClientGraphToDiagram interface {
+	Do(graph DiagramGraph) (ResponseDiagram, error)
 }
 
-func append3bytes(e, n, t byte) []byte {
-	c1 := e >> 2
-	c2 := (3&e)<<4 | n>>4
-	c3 := (15&n)<<2 | t>>6
-	c4 := 63 & t
-
-	var buf bytes.Buffer
-
-	buf.WriteByte(encode6bit(c1 & 63))
-	buf.WriteByte(encode6bit(c2 & 63))
-	buf.WriteByte(encode6bit(c3 & 63))
-	buf.WriteByte(encode6bit(c4 & 63))
-
-	return buf.Bytes()
-}
-
-func encode6bit(e byte) byte {
-	if e < 10 {
-		return 48 + e
-	}
-
-	e -= 10
-	if e < 26 {
-		return 65 + e
-	}
-
-	e -= 26
-	if e < 26 {
-		return 97 + e
-	}
-
-	e -= 26
-	switch e {
-	case 0:
-		return '-'
-	case 1:
-		return '_'
-	default:
-		return '?'
-	}
+// HttpClient http base client.
+type HttpClient interface {
+	Do(req *http.Request) (resp *http.Response, err error)
```

Comment:
**question** do we really need to have an abstraction for the `http.Client`?

Reply:
Yes ;)

Reply:
Reasons:
- Testability
- Flexibility to modify transport layer
- Flexibility for middleware
- General recommendation to make transport layer exchengable